### ISSUE TITLE
Mario Artist Paint Studio - Blinking with render frame buffer as texture

### ISSUE THREAD

**Issue by LuigiBlood** [gonetz/GLideN64#1156]
In 3D World mode (3rd option in the main menu) with "Render frame buffer as texture", the screen blinks a lot in black, sometimes more, sometimes less depending on the chosen world.

It also happens with the help screen that appears by default in animation mode (2nd option in main menu).

Should be known that it is required to use **Render frame buffer as texture** else the cursor does not render (not practical to select things).

I use this build on Project64: https://github.com/gonetz/GLideN64/issues/885#issuecomment-250971964

**Comment by LuigiBlood** [comment]
Basically a checkup: Still happens on 2.0 release.

**Comment by LuigiBlood** [comment]
Checkup: Still happening on d8ac5a7 WIP build.

**Comment by fzurita** [comment]
I think I can tell what is happening. Whenever it blinks, everything goes black but the hand pointer is the correct color.  It almost tells me that alpha is not being applied correctly for everything that is not the hand.

It could also tell me that we are not copying from the GPU memory to the CPU memory at the right time, so when the CPU renders the hand on top of the copied GPU buffer, it's copying over a black buffer instead.

**Comment by fzurita** [comment]
If you set "Copy video card frame buffer to N64 memory" to never, the blinking goes away.

**Comment by fzurita** [comment]
I have a theory. We are only copying the frame buffer to RDRAM when the frame changes. But the CPU is utilizing this data much more often than that. To test my theory, I will copy color buffer to RDRAM at each VI even though the FPS can be less than that. When we don't have a frame, I will use the previous's frame data.

**Comment by fzurita** [comment]
@gonetz when we do ```glReadPixels``` and copy the resulting buffer to RDRAM, we always take whatever alpha OpenGL gives us, is this correct behavior?

**Comment by gonetz** [comment]
It is complex question, which alpha should be stored in color buffer. You may check sources of angrylion's plugin, the places where the plugin writes processed pixel to the buffer. We currently can't do all calculations on pixel alpha, which actual RDP does. If you can suggest more precise way to copy data from OpenGL buffer to RDRAM, lets try it.

**Comment by fzurita** [comment]
I found ```static void fbwrite_32(
    uint32_t curpixel, uint32_t r, uint32_t g, uint32_t b, uint32_t blend_en,
    uint32_t curpixel_cvg, uint32_t curpixel_memcvg)
{
    uint32_t color;
    int coverage;
    unsigned long addr  = fb_address + 4*curpixel;
    addr &= 0x00FFFFFF;
    addr  = addr >> 2;

    coverage = finalize_spanalpha(blend_en, curpixel_cvg, curpixel_memcvg);
    color  = (r << 24) | (g << 16) | (b <<  8);
    color |= (coverage << 5);

    g = -(signed)(g & 1) & 3;
    PAIRWRITE32(addr, color, g, 0);
}```

in angrylion. Alpha written to RDRAM seems to depend on a coverage calculation.

**Comment by gonetz** [comment]
> Angrylion seems to be writing directly to RDRAM as it's calculating the color of each pixel.

Yes, Angrylion's plugin writes directly to RDRAM, but how it changes the fact that written alpha depends on coverage?

**Comment by fzurita** [comment]
It doesn't change that. It was just an observation.

**Comment by fzurita** [comment]
So this problem is probably related to this... we are copying the color buffer to RDRAM at the frame rate, so not necessarily at 60fps. We are copying RDRAM to color buffer at exactly 60FPS. We are copying RDRAM to color buffer that often probably because the cursor is updated at that rate.

**Comment by fzurita** [comment]
@gonetz I think clearing RDRAM only makes sense if we are not copying color buffer to RDRAM ourselves. Do you agree? Was the purpose of clearing RDRAM so that we could enable RDRAM to color buffer without necessarily enabling color buffer to RDRAM?

**Comment by LuigiBlood** [comment]
I had moments with some odd behavior on angrylion at times, where Project64 says it's at 40 VI/s but it's actually full speed.

**Comment by gonetz** [comment]
> Was the purpose of clearing RDRAM so that we could enable RDRAM to color buffer without necessarily enabling color buffer to RDRAM?

How it works on N64: 
 - RDP renders frame.
 - CPU renders something over that frame. Usually some small details.
-  VI displaying the resulted frame.

How it works with the plugin: 
 - plugin renders frame.
 - CPU renders something to corresponding frame in RDRAM.
-  VI blends frame in GPU with image from RDRAM and displaying the resulted frame.

We don't want to blend our hi-res frame with low-res image from RDRAM, but we have to because otherwise we will lost changes made by CPU. Thus, we clear frame in RDRAM after the blending, so next time it will contain only result of CPU rendering and our hi-res picture will not suffer.

The best way is to use FBInfo. In that case emulator will tell plugin, which pixels altered by CPU. Plugin can store these changes and apply them to hi-res image when necessary. No need to clear RDRAM buffer in that case.

**Comment by LuigiBlood** [comment]
Then the blending does not work on Paint Studio at all if I cannot see the cursor.

**Comment by LuigiBlood** [comment]
That said, shouldn't that code be disabled if we use render frame buffer as texture?

**Comment by AmbientMalice** [comment]
>Then the blending does not work on Paint Studio at all if I cannot see the cursor.

Are you referring to using FBInfo? Because it doesn't work currently.

**Comment by LuigiBlood** [comment]
Blending only works with FBInfo?

**Comment by AmbientMalice** [comment]
>Blending only works with FBInfo?

No, I'm saying that FBInfo doesn't work. FBInfo is frambuffer notifications. When the CPU modifies the FB, those changed pixels are copied, and nothing else. It's the ideal solution if it can be fixed. It would solve the problem of low resolution framebuffers needing to be copied on top of the (often high resolution) output, which basically forces native resolution.

**Comment by LuigiBlood** [comment]
Well Project64 doesn't support FBInfo, that I know. But in the meantime, I'm surprised nothing can be done about it.

But I do think if there's render frame buffer as texture enabled then that code is literally making more problems than it should have.

**Comment by AmbientMalice** [comment]
>But in the meantime, I'm surprised nothing can be done about it.

Mupen64plus was based on a version of mupen64 where FBInfo was broken. Logan made a PR to restore the functionality, but some adjustments need to be made on GLideN64's side, I believe. https://github.com/mupen64plus/mupen64plus-core/pull/376

>But I do think if there's render frame buffer as texture enabled then that code is literally making more problems than it should have.

Very likely so, yes.

**Comment by fzurita** [comment]
@gonetz do you see the problem in what we are doing if the VI/s doesn't match how often ```gDPFullSync``` is called and we are using RDRAM as our color buffer?

We will end up outputting a blank frame with only the CPU modifications on the frames where ```gDPFullSync``` was not called.

**Comment by gonetz** [comment]
> We will end up outputting a blank frame with only the CPU modifications on the frames where gDPFullSync was not called.

From the first glance you are right. 

> Also, why set everything to zero if we will overwrite all that when we copy the color buffer to RDRAM?

Copy color buffer to RDRAM usually disabled when we enable  copy color bufer from RDRAM. Otherwise we will get low res frame from RDRAM blended with our hires frame. Both options used together only in very special cases. For example, if we want to see monochrome "It's war!" cutscene in Conker.

**Comment by fzurita** [comment]
Ok, so special case with both enabled together is needed? Is how I did the pull request acceptable?

**Comment by fzurita** [comment]
Also, that does bring up the question, why is the menu all black if we don't enable copy color buffer to RDRAM? The cursor should work without it since the black should just be full alpha.

**Comment by gonetz** [comment]
< Ok, so special case with both enabled together is needed? Is how I did the pull request acceptable?

I haven't look at the PR yet. I need time to understand the problem it fixes and the path itself.  I did not check this particular ticket yet, I'm not even sure that I have the rom to run.

> the black should just be full alpha.

? May be I'm  missing context but black is zero.

**Comment by fzurita** [comment]
That's what I mean. when color buffer to RDRAM is off:
* When in the main menu of the game and RDRAM to color buffer is **enabled**, everything looks black with only a cursor showing.
* When in the main menu of the game and RDRAM to color buffer is **disabled**, everything looks normal, but the cursor is missing (the cursor is drawn by the CPU).

When we have RDRAM to color buffer enabled, the RDRAM buffer that we overlay over the color buffer should be fully transparent except for the cursor, but it's not, it's opaque, with a CPU drawn cursor.

**Comment by gonetz** [comment]
Sometimes  (or somegames ), CPU modifies all color buffer by setting alpha bit to 1 for all pixels in buffer.  I don't remember, where I saw this behavior,  but it is surely possible. When plugin checks color buffer validity by context, it applies mask to compared pixels, so if pixels difference only in alpha, they considered as equal. Thus, in this case we can have fully opaque color buffer in RDRAM even after we cleared it to zero on plugin's side.
